src-built-in/components/UIAPI/reducers/linker.ts

```diff
@@ -0,0 +1,184 @@
+
+    return new Promise((res, rej) => {
+        FSBL.Clients.LinkerClient.getAllChannels().forEach((channel: Channel) => {
+            initialChannels[nextChannelId] = {
+                id: nextChannelId,
+                name: channel.name,
+                color: channel.color,
+                active: false
+            };
+            initialNametoId[channel.name] = nextChannelId;
+            nextChannelId += 1;
+        });
+        linkerInitState.channels = initialChannels;
+        linkerInitState.nameToId = initialNametoId;
+
+        FSBL.Clients.RouterClient.addResponder("Finsemble.LinkerWindow.SetActiveChannels", function (err: any, msg: any) {
+            if (err) {
+                return rej(`Failed to add Finsemble.LinkerWindow.SetActiveChannels Responder: ${err}`);
+            }
+            const activeChannels = msg.data.channels;
+
+            const activeChannelIds: any[number] = [];
+            activeChannels.forEach((channel: Channel) => {
+                activeChannelIds.push(linkerInitState.nameToId[channel.name]);
+            });
+            const updatedChannels: Channels = Object.assign({}, linkerInitState.channels);
+            activeChannelIds.forEach((channelId: number) => {
+                updatedChannels[channelId].active = true;
+            });
+
+            linkerInitState.windowIdentifier = msg.data.windowIdentifier;
+            FSBL.Clients.Logger.system.log("toggle Linker window");
+            msg.sendQueryResponse(null, {});
+            store.dispatch(updateActiveChannels(msg.data));
+            
+            FSBL.Clients.ConfigClient.getValue("finsemble.accessibleLinker", (err: any, value: boolean) => {
+                if (err) {
+                    rej(`Error getting accessibleLinker value: ${err}`);
+                }
+                const newLinkerState = {
+                    ...linkerInitState,
+                    channels: updatedChannels,
+                    isAccessibleLinker: value,
+                };
+                res(newLinkerState);
+            });
+        });
+    });
+}
+
+function cleanUpAfterComponentUnmount() {
+    finsembleWindow.removeEventListener("blurred", () => {
+        finsembleWindow.hide();
+    });
+    finsembleWindow.removeEventListener("shown", () => {
+		finsembleWindow.focus();
+	});
+}
+
+// The linker's reducer
+const linker = (state = initialState, action: LinkerAction) => {
+    const { type, payload } = action;
+    switch (type) {
+        case LINKER_INIT:            
+            return loop(state, Cmd.run(initializeLinker, {
+                successActionCreator: initSuccess,
+            }));
+        case LINKER_INIT_SUCCESS:
+            const newState = payload.value;
+            return loop(newState, Cmd.run(() => FSBL.Clients.WindowClient.fitToDOM()));
+        case TOGGLE_CHANNEL_REQUEST:
+            const newState_request: Linker = {
+                ...state,
+                processingRequest: true
```

Comment:
@stephen-ciq @mesner I'm not sold on the extra code necessary to maintain a "processingRequest" state.

Reply:
It presupposed that a linker couldn't have two active requests. I could easily see a situation where an end user would toggle more than one color at a time.

Reply:
Good point.  @zona I think I agree with Terry.  It is a judgment call, but Terry and Ian have been clear in their design requirement for context-dependent actions.  So because of that, for things like linking which should happen very quickly, I think it's totally fine to not have a "processingRequest" flag in the state and explicitly document why.

Reply:
and also, maintaining the processingRequest for each link action is more work as Terry points out, so that's another reason to just not do it at all, in order to make the code easier to read.  the downside is that it's technically possibly for race conditions to occur, but we are explicitly making that tradeoff.

Reply:
I vote to keep it. Forces us to talk about the failure state.